pkg/operator/controller/controller.go

```diff
@@ -212,24 +214,32 @@ func haTypeForInfra(infraConfig *configv1.Infrastructure) ingressv1alpha1.Cluste
 	return ingressv1alpha1.UserDefinedClusterIngressHA
 }
 
-// effectiveHighAvailability returns the HA configuration as derived from the
-// given clusteringress and infrastructure config.
-func effectiveHighAvailability(ci *ingressv1alpha1.ClusterIngress, infraConfig *configv1.Infrastructure) (ha ingressv1alpha1.ClusterIngressHighAvailability) {
+// enforceEffectiveHighAvailability determines the effective HA configuration
+// for the given clusteringress and infrastructure config and publishes it to
+// the clusteringress's status.
+func (r *reconciler) enforceEffectiveHighAvailability(ci *ingressv1alpha1.ClusterIngress, infraConfig *configv1.Infrastructure) error {
+	// The clusteringress's HA configuration is immutable, so if we have
+	// published an HA type status, we must continue using it.
+	if len(ci.Status.HighAvailability.Type) > 0 {
+		return nil
+	}
+
 	switch {
 	case ci.Spec.HighAvailability != nil:
-		ha.Type = ci.Spec.HighAvailability.Type
+		ci.Status.HighAvailability.Type = ci.Spec.HighAvailability.Type
 	default:
-		ha.Type = haTypeForInfra(infraConfig)
+		ci.Status.HighAvailability.Type = haTypeForInfra(infraConfig)
 	}
-	return
+	if err := r.Client.Status().Update(context.TODO(), ci); err != nil {
```

Comment:
r.enforceEffectiveIngressDomain called Update on `ci` and now this is calling it on the same instance... in both cases you're mutating something that is almost certainly in an informer cache, which is a bug. I think you need to mutate copies before updating and then fetch the latest each time before proceeding.

Reply:
I agree we need to fetch the latest after updating so that we have the correct resource version in subsequent updates.  However, I'm confused about cache mutation.  We supply the empty `ClusterIngress`: https://github.com/openshift/cluster-ingress-operator/blob/dbb0ca8d8620eb761f6d0c48918fcf8b130dd092/pkg/operator/controller/controller.go#L95

So how can mutating it mutate the cache?

Reply:
Looks like the unstructure client copies the result into object that the caller (we) provided:
https://github.com/kubernetes-sigs/controller-runtime/blob/3285ad1e768092947e30e66155a44d3599502811/pkg/client/unstructured_client.go#L104

And actually, it _looks_ like we get back the updated object when we update the status—or at least I don't know what else we're getting back: https://github.com/kubernetes-sigs/controller-runtime/blob/3285ad1e768092947e30e66155a44d3599502811/vendor/k8s.io/client-go/dynamic/simple.go#L109-L117 https://github.com/kubernetes-sigs/controller-runtime/blob/3285ad1e768092947e30e66155a44d3599502811/pkg/client/unstructured_client.go#L143-L147

I wish this were documented...

Reply:
This statement seems to indicate there is no cache to mutate (emphasis added): "New returns a new Client using the provided config and Options. The returned client reads *and* writes directly from the server (**it doesn't use object caches**)." https://godoc.org/sigs.k8s.io/controller-runtime/pkg/client#New

Reply:
Yes it happens that we are passing around a non-caching client right now, but that's only because we're actively working around the controller-runtime defaults which would normally give us a client that reads from the underlying informer cache. And when reading from the caches (ultimately backed by `cache.Store`), mutating the cache objects violates the immutability contract and breaks thread-safety of the cache.

This is why I suggest we be extremely defensive and work with copies when doing *any mutations*. We're then guaranteed to be safe, and what do we lose? The extra spent cycles and allocation memory are clearly irrelevant in this context.

Reply:
Pushed a fix.

Reply:
Since we allocate the struct, if the client fetches from the cache, is our struct populated with references to fields from the cached object or do we get copied/unmarshalled stuff? Gotta did deeper I guess

Reply:
The autogenerated `DeepCopy` follows references and makes copies (e.g., if the original clusteringress's `.Spec.IngressDomain` is non-nil, `DeepCopy` makes a copy of the string and puts a reference to the new string in the copy clusteringresses's `.Spec.IngressDomain`; similar is done for other pointer values and values that contain references, such as label selectors).